File: java/src/test/java/benchmark/RoundtripTest.java
package benchmark;

import benchmark.model.Fixture;
import benchmark.model.v2.Generators;
import benchmark.serializers.BenchSerializer;
import benchmark.serializers.Registry;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertTrue;

class RoundtripTest {
  private static final String[] FIXTURES = {
    "message", "document", "telemetry", "strings", "event"
  };

  @Test
  void allSerializersRoundtripMessage() throws Exception {
    roundtripType("message");
  }

  @Test
  void allSerializersRoundtripAllFixtures() throws Exception {
    for (String type : FIXTURES) {
      roundtripType(type);
    }
  }

  private static void roundtripType(String type) throws Exception {
    Object expected = Generators.makeOne(type, Map.of(), 42L, 0);
    Fixture fx = new Fixture(type, expected);
    for (BenchSerializer ser : Registry.all()) {
      if (!ser.supports(type)) {
        continue;
      }
      ser.prepare(fx);
      byte[] buf = ser.serializeBytes(fx);
      Object out = ser.toDomain(ser.deserializeBytes(buf));
      assertTrue(
          Fidelity.check(expected, out),
          () -> "fidelity failed for " + ser.name() + " on " + type);
    }
  }

  @Test
  void compressSizesGzipHello() {
    int[] c = Compress.sizes("hello".getBytes(java.nio.charset.StandardCharsets.UTF_8));
    assertTrue(c[0] >= 20 && c[0] <= 40, "gzip=" + c[0]);
    int[] empty = Compress.sizes(new byte[0]);
    assertTrue(empty[0] == 0 && empty[1] == 0);
  }
}

File: kotlin/src/test/kotlin/benchmark/RoundtripTest.kt
package benchmark

import benchmark.model.Fixture
import benchmark.model.v2.Generators
import benchmark.serializers.Registry
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.nio.charset.StandardCharsets

class RoundtripTest {
    @Test
    fun allSerializersRoundtripMessage() {
        roundtripType("message")
    }

    @Test
    fun allSerializersRoundtripAllFixtures() {
        for (type in FIXTURES) {
            roundtripType(type)
        }
    }

    @Test
    fun compressSizesGzipHello() {
        val c = Compress.sizes("hello".toByteArray(StandardCharsets.UTF_8))
        assertTrue(c[0] in 20..40, "gzip=${c[0]}")
        val empty = Compress.sizes(ByteArray(0))
        assertTrue(empty[0] == 0 && empty[1] == 0)
    }

    companion object {
        private val FIXTURES = arrayOf("message", "document", "telemetry", "strings", "event")

        private fun roundtripType(type: String) {
            val expected = Generators.makeOne(type, emptyMap(), 42L, 0)
            val fx = Fixture(type, expected)
            for (ser in Registry.all()) {
                if (!ser.supports(type)) continue
                ser.prepare(fx)
                val buf = ser.serializeBytes(fx)
                val out = ser.toDomain(ser.deserializeBytes(buf))
                assertTrue(Fidelity.check(expected, out)) {
                    "fidelity failed for ${ser.name()} on $type"
                }
            }
        }
    }
}
